arch/xtensa/src/esp32/esp32_qencoder.c

```diff
@@ -0,0 +1,928 @@
+  regval |= PCNT_MODE_REVERSE << PCNT_CH0_LCTRL_MODE_U0_S; /* Rising A with B in HIGH = CW step */
+  regval |= PCNT_MODE_KEEP << PCNT_CH0_HCTRL_MODE_U0_S;    /* Rising A with B in LOW = CCW step */
+
+  putreg32(regval, PCNT_CONF0_U(priv->config->pcntid));
+
+  regval |= PCNT_COUNT_DEC << PCNT_CH1_NEG_MODE_U0_S; /* Ignore Falling-Edge */
+  regval |= PCNT_COUNT_INC << PCNT_CH1_POS_MODE_U0_S; /* Increase on Rising-Edge */
+
+  putreg32(regval, PCNT_CONF0_U(priv->config->pcntid));
+
+  /* Configure GPIO pins as Input with Pull-Up enabled */
+
+  esp32_configgpio(priv->config->ch0_gpio, INPUT_FUNCTION_3 | PULLUP);
+  esp32_configgpio(priv->config->ch1_gpio, INPUT_FUNCTION_3 | PULLUP);
+
+  /* Connect Channel A (ch0_gpio) and Channel B (ch1_gpio) crossed for X4 */
+
+  esp32_gpio_matrix_in(priv->config->ch0_gpio,
+                       priv->config->ch0_pulse_sig, 0);
+  esp32_gpio_matrix_in(priv->config->ch1_gpio,
+                       priv->config->ch0_ctrl_sig, 0);
+
+  esp32_gpio_matrix_in(priv->config->ch1_gpio,
+                       priv->config->ch1_pulse_sig, 0);
+  esp32_gpio_matrix_in(priv->config->ch0_gpio,
+                       priv->config->ch1_ctrl_sig, 0);
+
+  /* Clear the Reset bit to nable the Pulse Counter */
+
+  regval = getreg32(PCNT_CTRL_REG);
+  regval &= ~(1 << (2 * priv->config->pcntid));
+  putreg32(regval, PCNT_CTRL_REG);
+
+  esp32_dumpregs(priv, "After setup");
+
+  spin_unlock_irqrestore(&priv->lock, flags);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp32_shutdown
+ *
+ * Description:
+ *   This method is called when the driver is closed.  The lower half driver
+ *   should stop data collection, free any resources, disable timer hardware,
+ *   and put the system into the lowest possible power usage state
+ *
+ ****************************************************************************/
+
+static int esp32_shutdown(struct qe_lowerhalf_s *lower)
+{
+  struct esp32_lowerhalf_s *priv = (struct esp32_lowerhalf_s *)lower;
+
+  /* Disable PCNT clock */
+
+  modifyreg32(DPORT_PERIP_CLK_EN_REG, DPORT_PCNT_CLK_EN, 0);
+
+  /* Make sure initial position is 0 */
+
+  priv->position = 0;
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp32_position
+ *
+ * Description:
+ *   Return the current position measurement.
+ *
+ ****************************************************************************/
+
+static int esp32_position(struct qe_lowerhalf_s *lower, int32_t *pos)
+{
+  struct esp32_lowerhalf_s *priv = (struct esp32_lowerhalf_s *)lower;
+  irqstate_t flags;
+  int32_t position;
+  int16_t count;
+
+  DEBUGASSERT(lower && priv->inuse);
+
+  flags = spin_lock_irqsave(&priv->lock);
+
+  position = priv->position;
+  count = (int16_t) (getreg32(PCNT_CNT_U(priv->config->pcntid)) & 0xffff);
+
+  /* Return the position measurement */
+
+  *pos = position + (int32_t) count;
```

Comment:
```suggestion
  *pos = position + count;
```
not needed singe `position` is `int32_t` and `count` will be promoted automatically.